ruby/tests/basic_proto2.rb

```diff
@@ -0,0 +1,1261 @@
+      optional :optional_int64,  :int64,        2,  default: 2
+      optional :optional_uint32, :uint32,       3,  default: 3
+      optional :optional_uint64, :uint64,       4,  default: 4
+      optional :optional_bool,   :bool,         5,  default: true
+      optional :optional_float,  :float,        6,  default: 6
+      optional :optional_double, :double,       7,  default: 7
+      optional :optional_string, :string,       8,  default: "Default Str"
+      optional :optional_bytes,  :bytes,        9,  default: "\xCF\xA5s\xBD\xBA\xE6".force_encoding("ASCII-8BIT")
+      optional :optional_enum,   :enum,         10, "TestNonZeroEnum", default: 2
+      optional :optional_msg,    :message,      11, "TestMessage2"
+    end
+
+    add_message "TestMessage2" do
+      optional :foo, :int32, 1
+    end
+
+    add_enum "TestEnum" do
+      value :Default, 0
+      value :A, 1
+      value :B, 2
+      value :C, 3
+    end
+
+    add_enum "TestNonZeroEnum" do
+      value :A, 1
+      value :B, 2
+      value :C, 3
+    end
+
+    add_message "Recursive1" do
+      optional :foo, :message, 1, "Recursive2"
+    end
+    add_message "Recursive2" do
+      optional :foo, :message, 1, "Recursive1"
+    end
+
+    add_message "BadFieldNames" do
+      optional :dup, :int32, 1
+      optional :class, :int32, 2
+      optional :"a.b", :int32, 3
+    end
+    add_message "MapMessageWireEquiv" do
+      repeated :map_string_int32, :message, 1, "MapMessageWireEquiv_entry1"
+      repeated :map_string_msg, :message, 2, "MapMessageWireEquiv_entry2"
+    end
+    add_message "MapMessageWireEquiv_entry1" do
+      optional :key, :string, 1
+      optional :value, :int32, 2
+    end
+    add_message "MapMessageWireEquiv_entry2" do
+      optional :key, :string, 1
+      optional :value, :message, 2, "TestMessage2"
+    end
+
+    add_message "OneofMessage" do
+      oneof :my_oneof do
+        optional :a, :string, 1
+        optional :b, :int32, 2
+        optional :c, :message, 3, "TestMessage2"
+        optional :d, :enum, 4, "TestEnum"
+      end
+    end
+  end
+
+  Foo = pool.lookup("Foo").msgclass
+  Bar = pool.lookup("Bar").msgclass
+  Baz = pool.lookup("Baz").msgclass
+  TestMessage = pool.lookup("TestMessage").msgclass
+  TestMessage2 = pool.lookup("TestMessage2").msgclass
+  TestMessageDefaults = pool.lookup("TestMessageDefaults").msgclass
+  Recursive1 = pool.lookup("Recursive1").msgclass
+  Recursive2 = pool.lookup("Recursive2").msgclass
+  TestEnum = pool.lookup("TestEnum").enummodule
+  TestNonZeroEnum = pool.lookup("TestNonZeroEnum").enummodule
+  BadFieldNames = pool.lookup("BadFieldNames").msgclass
+  MapMessageWireEquiv = pool.lookup("MapMessageWireEquiv").msgclass
+  MapMessageWireEquiv_entry1 =
+    pool.lookup("MapMessageWireEquiv_entry1").msgclass
+  MapMessageWireEquiv_entry2 =
+    pool.lookup("MapMessageWireEquiv_entry2").msgclass
+  OneofMessage = pool.lookup("OneofMessage").msgclass
+
+# ------------ test cases ---------------
```

Comment:
There are a lot of tests same for proto3. Could you help refactor the existing test framework to make it work with both proto2 and proto3?